src/jobserver.h

```diff
@@ -0,0 +1,102 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+#define AUTH_KEY  "--jobserver-auth="
+#define FDS_KEY   "--jobserver-fds="
+#define FIFO_KEY  "fifo:"
+
+/// The GNU jobserver limits parallelism by assigning a token from an external
+/// pool for each command. On posix systems, the pool is a fifo or simple pipe
+/// with N characters. On windows systems, the pool is a semaphore initialized
+/// to N. When a command is finished, the acquired token is released by writing
+/// it back to the fifo or pipe or by increasing the semaphore count.
+///
+/// The jobserver functionality is enabled by passing --jobserver-auth=<val>
+/// (previously --jobserver-fds=<val> in older versions of Make) in the MAKEFLAGS
+/// environment variable and creating the respective file descriptors or objects.
+/// On posix systems, <val> is 'fifo:<name>' or '<read_fd>,<write_fd>' for pipes.
+/// On windows systems, <val> is the name of the semaphore.
+///
+/// The class parses the MAKEFLAGS variable and opens the object handle if needed.
+/// Once enabled, Acquire() must be called to acquire a token from the pool.
+/// If a token is acquired, a new command can be started.
+/// Once the command is completed, Release() must be called to return a token.
+/// The token server does not care in which order a token is received.
+struct Jobserver {
+  /// Parse the MAKEFLAGS environment variable to receive the path / FDs / name
+  /// of the token pool, and open the handle to the pool if it is an object.
+  /// If a jobserver argument is found in the MAKEFLAGS environment variable,
+  /// and the handle is successfully opened, later calls to Enable() return true.
+  /// If a jobserver argument is found, but the handle fails to be opened,
+  /// the ninja process is aborted with an error.
+  Jobserver();
+
+  /// Before exiting Ninja, ensure that tokens are returned and handles closed.
+  ~Jobserver();
+
+  /// Return true if jobserver functionality is enabled and initialized.
+  bool Enabled() const;
+
+  /// Return current token count or initialization signal if negative.
+  int Tokens() const { return token_count_; }
+
+  /// Implementation-specific method to acquire a token from the external pool
+  /// which is called for all tokens but returns early for the first token.
+  /// This method is called every time Ninja needs to start a command process.
+  /// Return true on success (token acquired), and false on failure (no tokens
+  /// available). First call always succeeds. Ninja is aborted on read errors.
+  bool Acquire();
```

Comment:
Unfortunately, I think this interface needs work. The GNU Make specification states that [It’s important that when you release the job slot, you write back the same character you read. Don’t assume that all tokens are the same character;](https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html), which means this method must at least return a token job byte value, that should be passed to Release().

Reply:
Technically Ninja is C++11 only, so no `std::optional<char>` is possible, but it would be easy to have a dedicated class such as:

```
struct Jobserver {
  ...
  struct Token {
    bool has_value() const { return value_ >= 0; }
    unsigned char value() const {
      assert(has_value() && "Cannot take value from invalid token");
      return static_cast<unsigned char>(value_);
    }
   private:
    int char value_ = -1;
  };
};
```

Which is easier to reason about than a magic integer.

Reply:
> Unfortunately, I think this interface needs work. The GNU Make specification states that 
> > It’s important that when you release the job slot, you write back the same character you read. Don’t assume that all tokens are the same character

Another thing I wish the manual didn't say...

This was discussed a little bit in #2450. As far as I can tell, there are no signs that this part of the jobserver design has ever or will ever be implemented anywhere. It may have been part of the original draft of how the jobserver will work, but then they realized they lacked a reason for one token to mean something different that another one. Current GNU Make source code uses the `+` and even mentions that it is ambiguous since success is determined by the read or write itself and not the content.

```
/* Token written to the pipe (could be any character...)  */
static char token = '+';
```

I think that it should be very easy to implement in the future if we *really* need to, so better to forget about it. However, if it makes you more comfortable of course I can do it now.

Reply:
I worked a little on saving the value of the token itself, and it actually allows for a little bit of simplification in some spots for existing files, so after I make sure it works right I might as well include it here. However, it makes the definition of the new `Clear()` function a little more complex...